### PR TITLE
[FIX] *: Records visible in demo not in data

### PR DESCRIPTION
**Pull request by jaeschwa**
Update all industry modules so all is_published are set to false in data and set to true in demo. The records are now only visible in demo and no more in data 

task-3945272

### CODE REVIEW

**Review by vava-odoo — CHANGES_REQUESTED** [review]
@jaeschwa I think you forgot half of the task description: 
- not published in data :heavy_check_mark:
- published in demo :negative_squared_cross_mark:

File: architects/demo/knowledge_article.xml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<?xml version='1.0' encoding='UTF-8'?>
+<odoo>
+    <record id="knowledge_article_3" model="knowledge.article">
+        <field name="is_published" eval="True"/>
+    </record>
```
Comment (vava-odoo) [inline architects/demo/knowledge_article.xml:5]: I thought we reached the agreement that knowledge articles shouldn't be published at all.
```suggestion
```

File: art_craft/demo/product_template.xml [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    </record>
+    <record id="product_template_24" model="product.template" context="{'create_product_product': False}">
+        <field name="is_published" eval="True"/>
+    </record>
+    <record id="product_template_25" model="product.template" context="{'create_product_product': False}">
+        <field name="is_published" eval="True"/>
+    </record>
+    <record id="product_template_26" model="product.template" context="{'create_product_product': False}">
+        <field name="is_published" eval="True"/>
+    </record>
```
Comment (vava-odoo) [inline art_craft/demo/product_template.xml:47]: Those are backend products that we don't want to publish at all.
```suggestion
```

File: art_craft/demo/product_template.xml [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    <record id="product_template_6" model="product.template" context="{'create_product_product': False}">
+        <field name="is_published" eval="True"/>
+    </record>
```
Comment (vava-odoo) [inline art_craft/demo/product_template.xml:50]:
```suggestion
```

File: art_craft/demo/product_template.xml [hunk 4]
```diff
@@ -0,0 +1,57 @@
+<?xml version='1.0' encoding='UTF-8'?>
+<odoo noupdate="1">
+    <record id="product_template_10" model="product.template" context="{'create_product_product': False}">
```
Comment (vava-odoo) [inline art_craft/demo/product_template.xml:3]: The context is necessary at record creation, but is not at record update.
```suggestion
    <record id="product_template_10" model="product.template">
```

File: art_craft/demo/product_template.xml [hunk 5]
```diff
@@ -0,0 +1,57 @@
+<?xml version='1.0' encoding='UTF-8'?>
+<odoo noupdate="1">
```
Comment (vava-odoo) [inline art_craft/demo/product_template.xml:2]: Did you test this? Because this file is not in the manifest, so never converted into the database...

File: bar_and_lounge/demo/product_product.xml [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    </record>
+    <record id="product_product_8" model="product.product">
+        <field name="is_published" eval="True"/>
+    </record>
+    <record id="product_product_9" model="product.product">
+        <field name="is_published" eval="True"/>
+    </record>
+    <record id="product_product_12" model="product.product">
+        <field name="is_published" eval="True"/>
+    </record>
```
Comment (vava-odoo) [inline bar_and_lounge/demo/product_product.xml:38]: No need to edit twice the same record
```suggestion
```

File: bar_and_lounge/demo/product_product.xml [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    <record id="product_product_14" model="product.product">
+        <field name="is_published" eval="True"/>
+    </record>
+    <record id="product_product_15" model="product.product">
+        <field name="is_published" eval="True"/>
+    </record>
+    <record id="product_product_16" model="product.product">
+        <field name="is_published" eval="True"/>
+    </record>
```
Comment (vava-odoo) [inline bar_and_lounge/demo/product_product.xml:20]: Those products are not for sale
```suggestion
```

File: cake_shop/demo/product_product.xml [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    </record>
+    <record id="product_product_57" model="product.product">
+        <field name="is_published" eval="True"/>
+    </record>
+    <record id="product_product_58" model="product.product">
+        <field name="is_published" eval="True"/>
+    </record>
+    <record id="product_product_61" model="product.product">
+        <field name="is_published" eval="True"/>
+    </record>
```
Comment (vava-odoo) [inline cake_shop/demo/product_product.xml:38]: Not for sale
```suggestion
```

File: clothing_boutique/demo/product_template.xml [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+    </record>
+    <record id="product_template_50" model="product.template" context="{'create_product_product': False}">
+        <field name="is_published" eval="True"/>
+    </record>
+    <record id="product_template_51" model="product.template" context="{'create_product_product': False}">
+        <field name="is_published" eval="True"/>
+    </record>
+    <record id="product_template_52" model="product.template" context="{'create_product_product': False}">
+        <field name="is_published" eval="True"/>
+    </record>
```
Comment (vava-odoo) [inline clothing_boutique/demo/product_template.xml:86]: Not for sale
```suggestion
```

File: clothing_boutique/demo/product_template.xml [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+    <record id="product_template_59" model="product.template" context="{'create_product_product': False}">
+        <field name="is_published" eval="True"/>
+    </record>
+    <record id="product_template_60" model="product.template" context="{'create_product_product': False}">
+        <field name="is_published" eval="True"/>
+    </record>
+    <record id="product_template_61" model="product.template" context="{'create_product_product': False}">
+        <field name="is_published" eval="True"/>
+    </record>
```
Comment (vava-odoo) [inline clothing_boutique/demo/product_template.xml:113]: Not for sale
```suggestion
```

**Review by vava-odoo — CHANGES_REQUESTED** [review]
Thanks for this @jaeschwa 

I'm stopping here the review, since I'm making several times the same comment. Could you apply the comments to all other occurrences?

It makes me think: we don't want this issue to occur again, so a generic test could help here.

Also don't forget the changes to do in the manifests when you add / remove a file.

Good luck with this :muscle: 

Cheers!

**Comment by vava-odoo** [comment]
Also the linked task is wrong, the correct ID is 3945272

File: sports_club/demo/appointment_type.xml [hunk 11]
```diff
@@ -0,0 +1,9 @@
+<?xml version='1.0' encoding='UTF-8'?>
+<odoo noupdate="1">
+    <record id="appointment_type_1" model="appointment.type">
+        <field name="is_published" eval="True"/>
+    </record>
+    <record id="appointment_type_2" model="appointment.type">
+        <field name="is_published" eval="True"/>
+    </record>
```
Comment (vava-odoo) [inline sports_club/demo/appointment_type.xml:8]: I think those ones do not need to be published, as it is accessible through an url link, constructed from the `short_code` of the `appointment.invite`

File: micro_brewery/demo/product_template.xml [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    </record>
+    <record id="product_template_23" model="product.template">
+        <field name="is_published" eval="True"/>
+    </record>
+    <record id="product_template_24" model="product.template">
+        <field name="is_published" eval="True"/>
+    </record>
+    <record id="product_template_29" model="product.template">
+        <field name="is_published" eval="True"/>
+    </record>
```
Comment (vava-odoo) [inline micro_brewery/demo/product_template.xml:38]:
```suggestion
```

File: micro_brewery/demo/product_template.xml [hunk 13]
```diff
@@ -0,0 +1,39 @@
+<?xml version='1.0' encoding='UTF-8'?>
+<odoo noupdate="1">
+    <record id="product_template_3" model="product.template">
+        <field name="is_published" eval="True"/>
+    </record>
+    <record id="product_template_2" model="product.template">
+        <field name="is_published" eval="True"/>
+    </record>
```
Comment (vava-odoo) [inline micro_brewery/demo/product_template.xml:8]: For a new file, I'd order it by id
```suggestion
    <record id="product_template_2" model="product.template">
        <field name="is_published" eval="True"/>
    </record>
    <record id="product_template_3" model="product.template">
        <field name="is_published" eval="True"/>
    </record>
```

File: industry_real_estate/__manifest__.py [hunk 14]
```diff
@@ -26,6 +26,7 @@
         'data/ir_model_2.xml',
         'data/crm_stages.xml',
         'data/res_config_setting.xml',
+        'data/sale_subscription_plan.xml',
```
Comment (vava-odoo) [inline industry_real_estate/__manifest__.py:29]: Well spotted! But the record defined inside is not used. I think it should replaced the one that is called on sale orders defined in demo (we should not rely on the ones from sale_subscription)

**Review by vava-odoo — COMMENTED** [review]
Great work! And tests will help future developments :ok_hand: 
I still have a few ( :eyes: ) comments, but keep the faith :muscle: 
Cheers

File: tests/test_generic/tests/test_manifest.py [hunk 15]
```diff
@@ -84,3 +89,27 @@ def _test_manifest_values(self, module, manifest_data):
...
+        data_folder = Path(get_industry_path() + module) / folder_name
+        if data_folder.exists():
+            # Get all files in the folder
+            data_files = {
+                str(file.relative_to(data_folder.parent))
+                for file in data_folder.glob('*')
+                if file.is_file()
+            }
+            # Get all files listed in the manifest for the folder
+            manifest_files = set(manifest_data.get(folder_name, []))
```
[the remark this answers is not in the document]
Comment (jaeschwa) [reply tests/test_generic/tests/test_manifest.py:103]: I used a set because the result are lists, and using a set will merge those list into a set, but i will add another test for the duplicates entries

File: tests/test_generic/tests/test_manifest.py [hunk 16]
```diff
@@ -84,3 +89,27 @@ def _test_manifest_values(self, module, manifest_data):
...
+            missing_in_manifest = data_files - manifest_files
+            for data_file in missing_in_manifest:
+                self.fail(
+                    f"File {data_file} in '{folder_name}' folder is not listed in the manifest."
+                )
+
+            # Check that all files listed in the manifest exist in the folder
+            missing_on_disk = manifest_files - data_files
+            for listed_file in missing_on_disk:
+                self.fail(f"Listed file {listed_file} in '{folder_name}' folder does not exist.")
```
[the remark this answers is not in the document]
Comment (jaeschwa) [reply tests/test_generic/tests/test_manifest.py:115]: Not in the testing you can remove files from the folder, the tests will pass. But you will have error later

File: tests/test_generic/tests/test_xml.py [hunk 17]
```diff
@@ -377,5 +410,50 @@ def _check_useless_fields_on_models(self, s, filename):
...
                         )
+
+    def _check_sale_published_conflict(self, xml_content, file_name):
+        if (
+            '<field name="sale_ok" eval="False"/>' in xml_content
+            and '<field name="is_published" eval="True"/>' in xml_content
+        ):
+            raise ValueError(
+                f"Conflict in {file_name}: 'sale_ok' cannot be False when 'is_published' is True."
+            )
```
[the remark this answers is not in the document]
Comment (jaeschwa) [reply tests/test_generic/tests/test_xml.py:425]: Clearly! The idea was to check as you said but with the complexity, I wrote a simpler test to ensure no inconsistencies.

File: tests/test_generic/tests/test_xml.py [hunk 18]
```diff
@@ -377,5 +410,50 @@ def _check_useless_fields_on_models(self, s, filename):
...
+
+    def _check_knowledge_article_is_published(self, xml_content, file_name):
+        if (
+            '<record ' in xml_content
+            and 'model="knowledge.article"' in xml_content
+            and '<field name="is_published" eval="True"/>' in xml_content
+        ):
+            raise ValueError(
+                f"Knowledge article in {file_name} should not have 'is_published' set to True."
+            )
```
[the remark this answers is not in the document]
Comment (jaeschwa) [reply tests/test_generic/tests/test_xml.py:435]: I will add testing in at the db level, these are simpler test to ensure no inconsistencies are found

File: sports_club/demo/product_template.xml [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+    </record>
+    <record id="product_template_17" model="product.template">
+        <field name="is_published" eval="True"/>
+    </record>
+    <record id="product_template_18" model="product.template">
+        <field name="is_published" eval="True"/>
+    </record>
+    <record id="product_template_20" model="product.template">
+        <field name="is_published" eval="True"/>
+    </record>
```
[the remark this answers is not in the document]
Comment (jaeschwa) [reply sports_club/demo/product_template.xml:23]: You are right cunsomable make no sense in the e-commerce.

File: tests/test_generic/tests/test_industry_requirements.py [hunk 20]
```diff
@@ -59,3 +73,20 @@ def test_cloc_exclude_view(self):
...
+        for model in models:
+            if "is_published" in self.env[model]._fields and "sale_ok" in self.env[model]._fields:
+                records = self.env[model].search(
+                    [("is_published", "=", True), ("sale_ok", "=", False)]
+                )
+                if records:
+                    record_info = ", ".join(
+                        ["%s (ID: %s)" % (record.display_name, record.id) for record in records]
+                    )
+                    self.assertFalse(
```
Comment (vava-odoo) [inline tests/test_generic/tests/test_industry_requirements.py:88]: Well, assertFalse after the if is useless. Replace the if by the assert and don't assign the info to an intermediate variable.

File: tests/test_generic/tests/test_xml.py [hunk 21]
```diff
@@ -377,5 +413,51 @@ def _check_useless_fields_on_models(self, s, filename):
...
                         )
+
+    def _check_sale_published_conflict(self, xml_content, file_name):
+        if (
+            '<field name="sale_ok" eval="False"/>' in xml_content
+            and '<field name="is_published" eval="True"/>' in xml_content
+        ):
+            _logger.warning(
+                f"Conflict in {file_name}: 'sale_ok' cannot be False when 'is_published' is True."
+            )
```
Comment (vava-odoo) [inline tests/test_generic/tests/test_xml.py:428]: To me, you can remove that test, as it is tested in the db. Also, as explained, you ensure with another test that `is_published` is defined in demo, while `sale_ok` will always be in data.

File: tests/test_generic/tests/test_xml.py [hunk 22]
```diff
@@ -377,5 +413,51 @@ def _check_useless_fields_on_models(self, s, filename):
...
+
+    def _check_knowledge_article_is_published(self, xml_content, file_name):
+        if (
+            '<record ' in xml_content
+            and 'model="knowledge.article"' in xml_content
+            and '<field name="is_published" eval="True"/>' in xml_content
+        ):
+            _logger.warning(
+                f"Knowledge article in {file_name} should not have 'is_published' set to True."
+            )
```
Comment (vava-odoo) [inline tests/test_generic/tests/test_xml.py:438]: Isn't this case handled with next test?

File: tests/test_generic/tests/test_xml.py [hunk 23]
```diff
@@ -377,5 +413,51 @@ def _check_useless_fields_on_models(self, s, filename):
...
+
+        try:
+            root = etree.fromstring(xml_content)
+            for record in root.xpath("//record"):
+                record_id = record.get("id")
+                model = record.get("model")
+                fields = frozenset(field.get("name") for field in record.xpath(".//field"))
+                record_key = (record_id, model)
+
+                if fields in records[record_key]:
```
Comment (vava-odoo) [inline tests/test_generic/tests/test_xml.py:456]: Shouldn't you check field by field (or any)?

**Review by vava-odoo — COMMENTED** [review]
Quite a big diff :black_flag: :black_heart: 
Almost there :muscle:

File: tests/test_generic/tests/test_xml.py (hunk 22 above)
Comment (jaeschwa) [reply tests/test_generic/tests/test_xml.py:438]: This test ensure that no knowledge.article are published in demo and data, the next test checks that no models in data is published

Comment (vava-odoo) [reply tests/test_generic/tests/test_xml.py:438]: Yeah, but since we don't edit knowledge article in demo... But ok, leave it there, it doesn't hurt

**Review by vava-odoo — APPROVED** [review]
